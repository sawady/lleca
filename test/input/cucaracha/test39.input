
fun factorial(n : Int) : Int {
    if n == 0 {
        res := 1
    } else {
        res := n * factorial(n - 1)
    }
    return res
}

fun main() {
    putNum(factorial(1)) putChar(10)
    putNum(factorial(2)) putChar(10)
    putNum(factorial(3)) putChar(10)
    putNum(factorial(4)) putChar(10)
    putNum(factorial(5)) putChar(10)
    putNum(factorial(6)) putChar(10)
    putNum(factorial(7)) putChar(10)
    putNum(factorial(8)) putChar(10)
    putNum(factorial(9)) putChar(10)
    putNum(factorial(10)) putChar(10)
}
