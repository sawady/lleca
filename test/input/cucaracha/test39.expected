Cons(
  Function(
    factorial,
    Cons(
      Parameter(
        n,
        Int
      ),
      Nil
    ),
    Int,
    Cons(
      StmtIf(
        ExprEq(
          ExprVar(
            n
          ),
          ExprConstNum(
            0
          )
        ),
        Cons(
          StmtAssign(
            res,
            ExprConstNum(
              1
            )
          ),
          Nil
        ),
        Just(
          Cons(
            StmtAssign(
              res,
              ExprMul(
                ExprVar(
                  n
                ),
                ExprCall(
                  factorial,
                  Cons(
                    ExprSub(
                      ExprVar(
                        n
                      ),
                      ExprConstNum(
                        1
                      )
                    ),
                    Nil
                  )
                )
              )
            ),
            Nil
          )
        )
      ),
      Cons(
        StmtReturn(
          ExprVar(
            res
          )
        ),
        Nil
      )
    )
  ),
  Cons(
    Function(
      main,
      Nil,
      Unit,
      Cons(
        StmtCall(
          putNum,
          Cons(
            ExprCall(
              factorial,
              Cons(
                ExprConstNum(
                  1
                ),
                Nil
              )
            ),
            Nil
          )
        ),
        Cons(
          StmtCall(
            putChar,
            Cons(
              ExprConstNum(
                10
              ),
              Nil
            )
          ),
          Cons(
            StmtCall(
              putNum,
              Cons(
                ExprCall(
                  factorial,
                  Cons(
                    ExprConstNum(
                      2
                    ),
                    Nil
                  )
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    10
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putNum,
                  Cons(
                    ExprCall(
                      factorial,
                      Cons(
                        ExprConstNum(
                          3
                        ),
                        Nil
                      )
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        10
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putNum,
                      Cons(
                        ExprCall(
                          factorial,
                          Cons(
                            ExprConstNum(
                              4
                            ),
                            Nil
                          )
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        putChar,
                        Cons(
                          ExprConstNum(
                            10
                          ),
                          Nil
                        )
                      ),
                      Cons(
                        StmtCall(
                          putNum,
                          Cons(
                            ExprCall(
                              factorial,
                              Cons(
                                ExprConstNum(
                                  5
                                ),
                                Nil
                              )
                            ),
                            Nil
                          )
                        ),
                        Cons(
                          StmtCall(
                            putChar,
                            Cons(
                              ExprConstNum(
                                10
                              ),
                              Nil
                            )
                          ),
                          Cons(
                            StmtCall(
                              putNum,
                              Cons(
                                ExprCall(
                                  factorial,
                                  Cons(
                                    ExprConstNum(
                                      6
                                    ),
                                    Nil
                                  )
                                ),
                                Nil
                              )
                            ),
                            Cons(
                              StmtCall(
                                putChar,
                                Cons(
                                  ExprConstNum(
                                    10
                                  ),
                                  Nil
                                )
                              ),
                              Cons(
                                StmtCall(
                                  putNum,
                                  Cons(
                                    ExprCall(
                                      factorial,
                                      Cons(
                                        ExprConstNum(
                                          7
                                        ),
                                        Nil
                                      )
                                    ),
                                    Nil
                                  )
                                ),
                                Cons(
                                  StmtCall(
                                    putChar,
                                    Cons(
                                      ExprConstNum(
                                        10
                                      ),
                                      Nil
                                    )
                                  ),
                                  Cons(
                                    StmtCall(
                                      putNum,
                                      Cons(
                                        ExprCall(
                                          factorial,
                                          Cons(
                                            ExprConstNum(
                                              8
                                            ),
                                            Nil
                                          )
                                        ),
                                        Nil
                                      )
                                    ),
                                    Cons(
                                      StmtCall(
                                        putChar,
                                        Cons(
                                          ExprConstNum(
                                            10
                                          ),
                                          Nil
                                        )
                                      ),
                                      Cons(
                                        StmtCall(
                                          putNum,
                                          Cons(
                                            ExprCall(
                                              factorial,
                                              Cons(
                                                ExprConstNum(
                                                  9
                                                ),
                                                Nil
                                              )
                                            ),
                                            Nil
                                          )
                                        ),
                                        Cons(
                                          StmtCall(
                                            putChar,
                                            Cons(
                                              ExprConstNum(
                                                10
                                              ),
                                              Nil
                                            )
                                          ),
                                          Cons(
                                            StmtCall(
                                              putNum,
                                              Cons(
                                                ExprCall(
                                                  factorial,
                                                  Cons(
                                                    ExprConstNum(
                                                      10
                                                    ),
                                                    Nil
                                                  )
                                                ),
                                                Nil
                                              )
                                            ),
                                            Cons(
                                              StmtCall(
                                                putChar,
                                                Cons(
                                                  ExprConstNum(
                                                    10
                                                  ),
                                                  Nil
                                                )
                                              ),
                                              Nil
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    ),
    Nil
  )
)
